File: queue.c
#include <stdio.h>
#include <stdlib.h>

struct Node {
    int data;
    struct Node* next;
};

void waitForEnter() {
    printf("\nPress Enter to continue...");
    int c;
    while ((c = getchar()) != '\n' && c != EOF); 
    getchar(); 
}

void enqueue(struct Node** front, struct Node** rear, int data, int showMessage);

void saveToFile(struct Node* front) {
    FILE *file = fopen("queue_data.txt", "w");
    if (file == NULL) return;
    
    struct Node* temp = front;
    while (temp != NULL) {
        fprintf(file, "%d\n", temp->data);
        temp = temp->next;
    }
    fclose(file);
}

void loadFromFile(struct Node** front, struct Node** rear) {
    FILE *file = fopen("queue_data.txt", "r");
    if (file == NULL) return; 
    
    int data;
    while (fscanf(file, "%d", &data) == 1) {
        enqueue(front, rear, data, 0); 
    }
    fclose(file);
}

void createQueue(struct Node** front, struct Node** rear) {
    if (*front != NULL) {
        printf("\n[Notice] A queue already exists! Please use Option 3 to Enqueue.\n");
        return;
    }

    char resp;
    int data;

    do {
        printf("Enter data to enqueue: ");
        
        if (scanf("%d", &data) != 1) {
            printf("  -> [Error] Invalid input! Letters/characters are not allowed.\n");
            printf("\nReturning to main menu...\n");
            return; 
        }
        
        enqueue(front, rear, data, 0); 

        while (1) {
            printf("Add another node [Y/N]? ");
            scanf(" %c", &resp); 
            
            if (resp == 'Y' || resp == 'y' || resp == 'N' || resp == 'n') {
                break; 
            } else {
                printf("  -> [Error] Invalid input. Please enter 'Y' for Yes or 'N' for No.\n");
            }
        }
        
    } while (resp == 'Y' || resp == 'y');
    
    printf("\n[Success] Initial Queue created!\n");
}

void display(struct Node** front, struct Node** rear) {
    struct Node* TravNode = *front;

    if (TravNode == NULL) {
        printf("\nError: The Queue is Empty!\n");
    } else {
        struct Node* tempFront = NULL;
        struct Node* tempRear = NULL;

        printf("\n--- Current Queue Line ---\n");
        printf("FRONT -> ");

        do {
            printf("[%d] -> ", TravNode->data);

            *front = TravNode->next;

            if (*front == NULL) {
                *rear = NULL;
            }

            TravNode->next = NULL;

            if (tempFront == NULL) {
                tempFront = tempRear = TravNode;
            } else {
                tempRear->next = TravNode;
                tempRear = TravNode;
            }

            TravNode = *front;

        } while (*front != NULL);

        printf("REAR\n");
        printf("--------------------------\n");

        *front = tempFront;
        *rear = tempRear;
    }
}

void enqueue(struct Node** front, struct Node** rear, int data, int showMessage) {
    struct Node* NewNode = (struct Node*)malloc(sizeof(struct Node));
    if (NewNode == NULL) {
        printf("\nError: Memory allocation failed!\n");
        return;
    }
    
    NewNode->data = data;
    NewNode->next = NULL;

    if (*rear == NULL) {
        *front = *rear = NewNode;
    } else {
        (*rear)->next = NewNode;
        *rear = NewNode;
    }
    
    if (showMessage) {
        printf("\n[Success] %d fell in line at the rear!\n", data);
    }
}

void dequeue(struct Node** front, struct Node** rear) {
    struct Node* DelNode = *front;

    if (DelNode == NULL) {
        printf("\nError: The Queue is Empty!\n");
        return;
    } 
    
    int servedData = DelNode->data; 
    
    *front = (*front)->next;
    DelNode->next = NULL;
    free(DelNode);
    
    if (*front == NULL) {
        *rear = NULL;
    }

    printf("\n[Success] %d was served and left the queue!\n", servedData);
}

void peekFront(struct Node* front) {
    if (front == NULL) {
        printf("\nError: The Queue is Empty!\n");
    } else {
        printf("\n[Peek] The next in line at the front is: %d\n", front->data);
    }
}

int main() {
    struct Node* front = NULL; 
    struct Node* rear = NULL;  
    int choice, data;

    loadFromFile(&front, &rear);

    while (1) {
        system("cls"); 
        
        printf("===== Queue Implementation by Group 1 =====\n");
        printf("1. Create queue\n");
        printf("2. Display queue\n");
        printf("3. Enqueue (insert item)\n");
        printf("4. Dequeue (remove item)\n");
        printf("5. Peek (view front item)\n");
        printf("6. Exit\n");
        printf("Enter your choice: ");
        
        if (scanf("%d", &choice) != 1) {
            printf("\n[Error] Invalid input! Letters/symbols are not allowed.\n");
            waitForEnter(); 
            continue; 
        }

        switch (choice) {
            case 1:
                createQueue(&front, &rear);
                waitForEnter();
                break;
            case 2:
                display(&front, &rear);
                waitForEnter();
                break;
            case 3:
                printf("Enter data to enqueue: ");
                if (scanf("%d", &data) != 1) {
                    printf("\n[Error] Invalid input! Please enter a number.\n");
                } else {
                    enqueue(&front, &rear, data, 1); 
                }
                waitForEnter();
                break;
            case 4:
                dequeue(&front, &rear);
                waitForEnter();
                break;
            case 5:
                peekFront(front);
                waitForEnter();
                break;
            case 6:
                printf("\nSaving data and exiting program...\n");
                saveToFile(front);
                
                while (front != NULL) {
                    struct Node* temp = front;
                    front = front->next;
                    free(temp);
                }
                exit(0);
            default:
                printf("\n[Error] Invalid selection.\n");
                waitForEnter();
        }
    }
    return 0;
}

File: queue.py
from os import system
from pathlib import Path

class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

def waitForEnter():
    print("\nPress Enter to continue...")
    input()

def saveToFile(front):
    with open("queue_data.txt", "w") as file:
        temp = front
        while temp != None:
            file.write(f"{temp.data}")
            temp = temp.next

def enqueue(front, rear, data, showMessage):
    newNode = Node(data)

    if rear == None:
        front = rear = newNode
    else:
        rear.next = newNode
        rear = newNode

    if showMessage:
        print(f"\n[Success] {data} fell in line at the rear!")

    return front, rear

def loadFromFile(front, rear):
    if not Path("queue_data.txt").exists():
        return front, rear

    with open("queue_data.txt", "r") as file:
        for line in file:
            data = line.strip()
            if data:
                front, rear = enqueue(front, rear, int(data), 0)

    return front, rear

def createQueue(front, rear):
    if front != None:
        print("\n[Notice] A queue already exists! Please use Option 3 to Enqueue.")
        return front, rear

    resp = ""
    data = 0

    while True:
        print("Enter data to enqueue: ")

        try:
            data = int(input())
        except:
            print("  -> [Error] Invalid input! Letters/characters are not allowed.")
            print("\nReturning to main menu...\n")
            return front, rear

        front, rear = enqueue(front, rear, data, 0)

        while True:
            print("Add another node [Y/N]? ")
            resp = input().strip()

            if resp in ['Y', 'y', 'N', 'n']:
                break
            else:
                print("  -> [Error] Invalid input. Please enter 'Y' for Yes or 'N' for No.")

        if resp not in ['Y', 'y']:
            break

    print("\n[Success] Initial Queue created!")
    return front, rear

def display(front, rear):
    travNode = front

    if travNode == None:
        print("\nError: The Queue is Empty!")
        return front, rear

    tempFront = None
    tempRear = None

    print("\n--- Current Queue Line ---")
    print("FRONT -> ", end="")

    while True:
        print(f"[{travNode.data}] -> ", end="")

        front = travNode.next

        if front == None:
            rear = None

        travNode.next = None

        if tempFront == None:
            tempFront = tempRear = travNode
        else:
            tempRear.next = travNode
            tempRear = travNode

        travNode = front

        if front == None:
            break

    print("REAR")
    print("--------------------------")

    return tempFront, tempRear

def dequeue(front, rear):
    delNode = front

    if delNode == None:
        print("\nError: The Queue is Empty!")
        return front, rear

    servedData = delNode.data

    front = front.next
    delNode.next = None

    if front == None:
        rear = None

    print(f"\n[Success] {servedData} was served and left the queue!")

    return front, rear

def peekFront(front):
    if front == None:
        print("\nError: The Queue is Empty!")
    else:
        print(f"\n[Peek] The next in line at the front is: {front.data}")

def main():
    front = None
    rear = None

    front, rear = loadFromFile(front, rear)

    while True:
        system("cls")

        print("===== Queue Implementation by Group 1 =====")
        print("1. Create queue")
        print("2. Display queue")
        print("3. Enqueue (insert item)")
        print("4. Dequeue (remove item)")
        print("5. Peek (view front item)")
        print("6. Exit")
        print("Enter your choice: ")

        try:
            choice = int(input())
        except:
            print("\n[Error] Invalid input! Letters/symbols are not allowed.")
            waitForEnter()
            continue

        if choice == 1:
            front, rear = createQueue(front, rear)
            waitForEnter()
        elif choice == 2:
            front, rear = display(front, rear)
            waitForEnter()
        elif choice == 3:
            print("Enter data to enqueue: ")
            try:
                data = int(input())
                front, rear = enqueue(front, rear, data, 1)
            except:
                print("\n[Error] Invalid input! Please enter a number.")
            waitForEnter()
        elif choice == 4:
            front, rear = dequeue(front, rear)
            waitForEnter()
        elif choice == 5:
            peekFront(front)
            waitForEnter()
        elif choice == 6:
            print("\nSaving data and exiting program...")
            saveToFile(front)

            while front != None:
                temp = front
                front = front.next

            exit(0)
        else:
            print("\n[Error] Invalid selection.")
            waitForEnter()

main()
